contracts/margin/external/BucketLender/BucketLender.sol

```diff
@@ -0,0 +1,817 @@
+        require(loanOffering.lenderFee == 0);
+        // no need to require anything about loanOffering.takerFee
+        // no need to require anything about loanOffering.expirationTimestamp
+        // no need to require anything about loanOffering.salt
+
+        /* CHECK VALUES32 */
+        // no need to require anything about loanOffering.callTimeLimit
+        // no need to require anything about loanOffering.maxDuration
+        // no need to require anything about loanOffering.interestRate
+        // no need to require anything about loanOffering.interestPeriod
+
+        return address(this);
+    }
+
+    /**
+     * Called by the Margin contract when anyone transfers ownership of a loan to this contract.
+     * This function initializes this contract and returns this address to indicate to Margin
+     * that it is willing to take ownership of the loan.
+     *
+     * @param  from        (unused)
+     * @param  positionId  Unique ID of the position
+     * @return             This address on success, throw otherwise
+     */
+    function receiveLoanOwnership(
+        address from,
+        bytes32 positionId
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        MarginCommon.Position memory position = MarginHelper.getPosition(DYDX_MARGIN, POSITION_ID);
+
+        assert(position.principal > 0);
+        assert(position.owedToken == OWED_TOKEN);
+        assert(position.heldToken == HELD_TOKEN);
+
+        // set relevant constants
+        uint256 initialPrincipal = position.principal;
+        principalForBkt[0] = initialPrincipal;
+        principalTotal = initialPrincipal;
+        weightForBkt[0] = weightForBkt[0].add(initialPrincipal);
+        weightForBktForAct[0][from] = weightForBktForAct[0][from].add(initialPrincipal);
+
+        return address(this);
+    }
+
+    /**
+     * Called by Margin when additional value is added onto the position this contract
+     * is lending for. Balance is added to the address that loaned the additional tokens.
+     *
+     * @param  payer           Address that loaned the additional tokens
+     * @param  positionId      Unique ID of the position
+     * @param  principalAdded  Amount that was added to the position
+     *  param  lentAmount      (unused)
+     * @return                 This address to accept, a different address to ask that contract
+     */
+    function increaseLoanOnBehalfOf(
+        address payer,
+        bytes32 positionId,
+        uint256 principalAdded,
+        uint256 lentAmount
+    )
+        external
+        onlyMargin
+        onlyPosition(positionId)
+        returns (address)
+    {
+        // Don't allow other lenders
+        require(payer == address(this));
+
+        // p2 is the principal after the add (p2 > p1)
+        // p1 is the principal before the add
+        uint256 principalAfterIncrease = getCurrentPrincipalFromMargin();
```

Comment:
[no-action] Ok this is correct, but I had to check. The implementation of `IncreasePositionImpl#updateState` does update the principal before calling out. Seems finnicky to rely on that assumption, but I guess we have to